hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/safemode/ContainerSafeModeRule.java

```diff
@@ -50,10 +57,14 @@ public class ContainerSafeModeRule extends
   // Required cutoff % for containers with at least 1 reported replica.
   private double safeModeCutoff;
   // Containers read from scm db (excluding containers in ALLOCATED state).
-  private Map<Long, ContainerInfo> containerMap;
-  private double maxContainer;
-
-  private AtomicLong containerWithMinReplicas = new AtomicLong(0);
+  private Map<Long, Integer> ratisContainerMinReplicaMap;
+  private Map<Long, Set<UUID>> ratisContainerDNsMap;
+  private Map<Long, Integer> ecContainerMinReplicaMap;
```

Comment:
We don't need to maintain this mapping either, as far as I can tell. Whenever we have the container id and need the replication factor of that container, we can use a container manager method to get it. That'll be a constant time (O(1)) lookup for container manager.

Reply:
This part of the logic has also been improved.